lib/maths/CBayesianOptimisation.cc

```diff
@@ -372,6 +398,268 @@ double CBayesianOptimisation::kernel(const TVector& a, const TVector& x, const T
+    try {
+
+        // Call name() to prime the traverser if it hasn't started
+        traverser.name();
+        if (traverser.isEof()) {
+            //            m_RestoredStateDetail.s_RestoredStateStatus = E_NoDetectorsRecovered;
+            LOG_ERROR(<< "Expected persisted state but no state exists");
+            return false;
+        }
+
+        do {
+            const std::string& name = traverser.name();
+            if (name == MIN_BOUNDARY_TAG) {
+                if (restoreSubLevelVector(BOUNDARY_TAG, "minimum boundary",
+                                          m_MinBoundary, traverser) == false) {
+                    return false;
+                }
+            } else if (name == MAX_BOUNDARY_TAG) {
+                if (restoreSubLevelVector(BOUNDARY_TAG, "maximum boundary",
+                                          m_MaxBoundary, traverser) == false) {
+                    return false;
+                }
+            } else if (name == ERROR_VARIANCES_TAG) {
+                if (restoreSubLevelVector(ERROR_VARIANCE_TAG, "error variance",
+                                          m_ErrorVariances, traverser) == false) {
+                    return false;
+                }
+            } else if (name == KERNEL_PARAMETERS_TAG) {
+                if (restoreSubLevelVector(KERNEL_PARAMETER_TAG, "kernel parameters",
+                                          m_KernelParameters, traverser) == false) {
+                    return false;
+                }
+            } else if (name == MIN_KERNEL_COORDINATE_DISTANCE_SCALES_TAG) {
+                if (restoreSubLevelVector(MIN_KERNEL_COORDINATE_DISTANCE_SCALE_TAG,
+                                          "minimum kernel coordinate distance scale",
+                                          m_MinimumKernelCoordinateDistanceScale,
+                                          traverser) == false) {
+                    return false;
+                }
+            } else if (name == FUNCTION_MEAN_VALUES_TAG) {
+                if (restoreFunctionMeanValues(m_FunctionMeanValues, traverser) == false) {
+                    return false;
+                }
+            } else {
+                LOG_ERROR(<< "Unexpected name for restoring bayesian optimization parameters: "
+                          << traverser.name());
+                return false;
+            }
+        } while (traverser.next());
+    } catch (std::exception& e) {
+        LOG_ERROR(<< "Failed to restore state! " << e.what());
+        return false;
+    }
+
+    return true;
+}
+
+bool CBayesianOptimisation::restoreSubLevelVector(const std::string& tag,
+                                                  const std::string& name,
+                                                  TVector& vector,
+                                                  core::CStateRestoreTraverser& traverser) {
+    if (traverser.hasSubLevel()) {
```

Comment:
In general, I'm not sure it is worth doing the checking for sublevels explicitly in the code: we don't in any other restore code and it bloats the code a bit. The error will already be detected by `traverser.traverseSubLevel(restoreVector(tag, vector))` which will return false. I know you could argue that it is useful to have the separate error message, but I think on balance it is overkill. From experience, if you get a problem with restore you really need to inspect the whole state document to work out what's wrong.

To me this points to just ditching these wrapper `restoreSubLevel*` functions.

Reply:
I removed the checks and `restoreSubLevel` functions in all places but one, where I need it to have a correct restoration behavior if no points were added to the optimizer yet.